### PR TITLE
Accordion

### PR DESCRIPTION
**Pull request by sagarjadhav**
### Adds Accordion Component
Screenshot 2023-11-22 at 4 57 58 PM

### CODE REVIEW

File: src/accordion/index.html [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+		}
+
+		tp-accordions-accordion {
+			margin-top: 30px;
+		}
+	</style>
+</head>
+<body>
+<main>
+	<tp-accordions expand-all="no" collapse-all="no">
```
Comment (junaidbhura) [inline src/accordion/index.html:26]: @sagarjadhav Lets rename to `tp-accordion`

File: src/accordion/index.html [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		<tp-accordions-expand-all>
```
Comment (junaidbhura) [inline src/accordion/index.html:27]: @sagarjadhav Lets rename to `tp-accordion-expand-all`

File: src/accordion/index.html [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+			<button>Expand All</button>
+		</tp-accordions-expand-all>
+		<tp-accordions-collapse-all>
```
Comment (junaidbhura) [inline src/accordion/index.html:30]: @sagarjadhav Lets rename to `tp-accordion-collapse-all`

File: src/accordion/index.html [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+			<button>Collapse All</button>
+		</tp-accordions-collapse-all>
+
+		<tp-accordions-accordion open="yes">
```
Comment (junaidbhura) [inline src/accordion/index.html:34]: @sagarjadhav Lets rename to `tp-accordion-item`

File: src/accordion/index.html [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+			<tp-accordions-accordion-handle>
```
Comment (junaidbhura) [inline src/accordion/index.html:35]: @sagarjadhav Lets rename to `tp-accordion-handle`

File: src/accordion/index.html [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+				<button>Accordion title</button>
+			</tp-accordions-accordion-handle>
+			<tp-accordions-accordion-content>
```
Comment (junaidbhura) [inline src/accordion/index.html:38]: @sagarjadhav Lets rename to `tp-accordion-content`

File: src/accordion/index.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+/**
+ * Styles.
+ */
+import './style.scss';
+
+/**
+ * Components.
+ */
+import { TPAccordionsElement } from './tp-accordions';
```
Comment (junaidbhura) [inline src/accordion/index.ts:9]: @sagarjadhav Lets rename these classes based on the updated names

File: src/accordion/tp-accordions-accordion-handle.ts [hunk 8]
```diff
@@ -0,0 +1,46 @@
+import { TPAccordionsElement } from './tp-accordions';
+
+/**
+ * TPAccordionsAccordionHandleElement.
+ */
+export class TPAccordionsAccordionHandleElement extends HTMLElement {
+	private accordions: TPAccordionsElement | null;
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion-handle.ts:7]: @sagarjadhav Lets remove this

File: src/accordion/tp-accordions-accordion-handle.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+	constructor() {
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion-handle.ts:8]: @sagarjadhav Lets remove this

File: src/accordion/tp-accordions-accordion-handle.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+		super();
+
+		this.accordions = this.closest( 'tp-accordions' );
+	}
+
+	connectedCallback() {
+		this.addEventListener( 'click', () => this.toggle() );
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion-handle.ts:15]: @sagarjadhav The click event should be on the button

File: src/accordion/tp-accordions-accordion-handle.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	 * Toggle accordion state.
+	 */
+	toggle() {
+		if ( this.accordions ) {
+			this.accordions.setAttribute( 'expand-all', 'no' );
+			this.accordions.setAttribute( 'collapse-all', 'no' );
+		}
+
+		// Toggle accordion item states.
+		this.parentElement?.toggleAttribute( 'open' );
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion-handle.ts:28]: @sagarjadhav Lets use `this.closest( 'tp-accordion' )`

File: src/accordion/tp-accordions-accordion-handle.ts [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	/**
+	 * Open accordion.
+	 */
+	open() {
+		// Set the attributes of the accordion to expand it.
+		this.setAttribute( 'open', 'yes' );
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion-handle.ts:36]: @sagarjadhav We should be setting an attribute on the parent

File: src/accordion/tp-accordions-accordion-handle.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+	}
+
+	/**
+	 * Close accordion.
+	 */
+	close() {
+		// Set the attributes of the accordion to collapse it.
+		this.removeAttribute( 'open' );
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion-handle.ts:44]: @sagarjadhav We should be setting an attribute on the parent

File: src/accordion/tp-accordions-accordion.ts [hunk 14]
```diff
@@ -0,0 +1,58 @@
+/**
+ * TpAccordionsAccordionElement.
+ */
+import { TPAccordionsAccordionContentElement } from './tp-accordions-accordion-content';
+import { slideElementDown, slideElementUp } from '../global/utility';
+
+export class TPAccordionsAccordionElement extends HTMLElement {
+	private content: TPAccordionsAccordionContentElement | null;
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion.ts:8]: @sagarjadhav Lets remove

File: src/accordion/tp-accordions-accordion.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+	constructor() {
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-accordion.ts:10]: @sagarjadhav Lets remove this

File: src/accordion/tp-accordions-collapse-all.ts [hunk 16]
```diff
@@ -0,0 +1,20 @@
+/**
+ * TPAccordionsCollapseAllElement.
+ */
+import { TPAccordionsElement } from './tp-accordions';
+
+export class TPAccordionsCollapseAllElement extends HTMLElement {
+	connectedCallback() {
+		this.addEventListener( 'click', () => this.collapseAllAccordions() );
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-collapse-all.ts:8]: @sagarjadhav The event listener should be on the button

File: src/accordion/tp-accordions-collapse-all.ts [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+	}
+
+	collapseAllAccordions() {
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-collapse-all.ts:11]: @sagarjadhav Lets rename to `collapseAll`

File: src/accordion/tp-accordions-collapse-all.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+		const accordions: TPAccordionsElement | null = this.closest( 'tp-accordions' );
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-collapse-all.ts:12]: @sagarjadhav From here, we should only be setting an attribute on the parent

File: src/accordion/tp-accordions-expand-all.ts [hunk 19]
```diff
@@ -0,0 +1,21 @@
+import { TPAccordionsElement } from './tp-accordions';
+
+/**
+ * TPAccordionsExpandAllElement.
+ */
+export class TPAccordionsExpandAllElement extends HTMLElement {
```
Comment (junaidbhura) [inline src/accordion/tp-accordions-expand-all.ts:6]: @sagarjadhav Same feedback on this component

**Review by junaidbhura — CHANGES_REQUESTED** [review]

File: src/accordion/tp-accordions-accordion-handle.ts (hunk 11 above)
Comment (imranhsayed) [reply src/accordion/tp-accordions-accordion-handle.ts:28]: I believe you mean `this.closest( 'tp-accordion-item' );` @junaidbhura .

File: src/accordion/tp-accordions-collapse-all.ts (hunk 18 above)
Comment (imranhsayed) [reply src/accordion/tp-accordions-collapse-all.ts]: @junaidbhura I believe we are setting to 'tp-accordions' here. So the parent controls the children.

**Review by junaidbhura — APPROVED** [review]

**Pull request merged**